BigAnimal - CSP tagging (#5863)
EnterpriseDB/docs: product_docs/docs/biganimal/release/using_cluster/managing_cloud_account/csp_tagging.mdx

---
title: "Tagging AWS resources"
deepToC: true
---

Tags are key-value pairs you can apply to cloud provider resources created on your own cloud account (Your Cloud Account). Resource tagging allows you to have fine-grained access control and granular cost data at the bucket level. When you add a tag, the following resources are labeled: 

- EC2 instances
- S3 buckets
- Load balancers
- Volumes

> You can create and manage tags on a per-project basis in the EDB Postgres® AI Console. The Console then applies the tag to the resources of all regions configured in your account. 

Reviewer: @amine-khemissi and @ivoryibuaka, I rephrased this part to make it more clear, let me know what you think. 

Cloud Service is the new name that the product team assigned to BigAnimal, and Console is the term to refer to the dashboard. I used Console here as I believe makes it more clear.

Reviewer 2: the phrasing is much better, thank you !

## Create Tags

To create a tag:

1.  Log in to the Console.
    
1.  Go to the project's page.

1.  In the project's list, select your **Project**.

1.  On your project's home page left navigation, select **Cloud Providers**.

1.  On the cloud provider's page, select **Your Cloud Account**.

1.  On your cloud account tab, under **AWS**, select **Manage Tags**.

1.  Select **Add a Key Value pair**.

1.  Provide **Key** and **Value** and select **Save**.

1.  A **Confirm changes to AWS tags** message pops up.

1.  Select **Confirm** to add the tag.

1.  This **Key:Value** pair is added at your project level. It is populated on the EC2 instances, S3 buckets, load balancers and volumes managed in that project, for all regions and clusters.

This new tag appears in the list under **AWS** on Cloud Service Providers. 

This tag is also propagated on your AWS resources. For instance, you can view this tag in the **Tags** tab of your EC2 instance resources on the AWS console.

> ### Considerations

Reviewer: @amine-khemissi I reworked this section, let me know what you think. I'd prefer keeping the word considerations instead of limitations, as I don't expect many customers to encounter these issues, and limitations have a negative connotation, but I will bring this up with the docs team. 

@navnathedb, let me know if there are other BA reserved words we should add here.

Reviewer 3: Along with reversed keys, this doc also needs to talk default tags applied on Cluster and Project resources

<img width="567" alt="Screenshot 2024-07-19 at 3 48 57 PM" src="https://github.com/user-attachments/assets/7fbe4d45-b2e7-4f59-b862-1bd3aa14c471">

ref: https://github.com/EnterpriseDB/upm-rfcs/blob/main/rfcs/rfc-upm-34349-csp-tag-controller.md#3-design

Reviewer: Hi @dhope-nagesh  
Is it these four tags only? Or might there be other BA-generated tags? And do these tags appear in the BA and AWS dashboards as well? Or are they "hidden"? 

We already mention the existence of the BA default tags in the context of the maximum limitation being 50 tags per resource (line 53). So now I am wondering if we should provide a list of all default BA tags? But I think that depends on 1. whether these are visible to them or not and 2. if they need to know the names of the tags for other operations.

Reviewer 3: These 4 tags are part of the requirement from the product team, but they are not visible on the BA portal and they visible on AWS.

cc @amine-khemissi @ivoryibuaka

Reviewer: It sounds like these tags are internal implementation details (required for inner AWS-BigAnimal workings), and not so much for the customer to do something with them. 

If that is the case, we can mention that these tags exist, but I don't believe we need to add a list/table with the tag names here. Let me know what you think :) 

WDYT?

Reviewer: Yes, I think I understand now. We don't need to list which tags come from BA by default, we just need to specify which prefixes and which keys are reserved, so they understand why they can't use them. I hope that makes sense?

In that case, I can add an expandable section with the list you provided above and a note saying these are the values they can't use.

Reviewer 2: that is a good idea @gvasquezvargas

Reviewer: Hi all, I have edited this section. Here is the preview so you can see how the expand is displayed. 
https://deploy-preview-5863--edb-docs-staging.netlify.app/docs/biganimal/latest/using_cluster/managing_cloud_account/csp_tagging/#considerations 

@navnathedb you mentioned you are going to put a link to docs from the UI? If yes, I have created a permanent URL redirect (PURL) that is more robust than the dynamic URL. The idea behind it is that, even if we make changes to the documentation and move this page around, there will always be a redirect in place to link to this content.  So for the the UI link we can use: https://www.enterprisedb.com/docs/purl/cloudservice/csp_tagging/#considerations 
Let me know if I can give you more information on this :). 
PD:  The redirect does not work right now, but it will work once we merge and publish this PR.

Reviewer 4: We need to add doc link on the tag validation if error occurs. 
Detailed discussion happened on [Slack](https://edb-starlight.slack.com/archives/C02K4TSUCRJ/p1721834254729869).

Consider following while adding a **Key:Value** pair:

> - Review the [tagging limitations of AWS](https://docs.aws.amazon.com/emr/latest/EMR-Serverless-UserGuide/tagging-restrictions.html).

Reviewer 4: This is the AWS EMR tagging limitations. AWS has different policies for applying tags to different resources, and we have chosen the most common policies that support all resources. So, I don't think adding only EMR policies will be good idea?

cc @amine-khemissi @dhope-nagesh

Reviewer: @navnathedb I think I found the correct AWS page to link towards: 

https://docs.aws.amazon.com/tag-editor/latest/userguide/tagging.html#tag-conventions 

Can you confirm?

Reviewer 2: +1

> - The **Key** cannot have `aws` or `managedby` in the prefix, as they are reserved by AWS and EDB Cloud Service. 

Reviewer 4: - `aws:` and `AWS:` both are the reserved prefixes(meaning user can't create tag with key`aws:Team`)
  - `ManagedBy` is the reserved tag key (user can create tag key with `ManagedByABC`)

> - In AWS, a resource can have a maximum number of 50 tags. The EDB Cloud Service applies some tags by default, so we recommend not adding more than 40 of your own tags. Take into account that the number of tags per resource is defined by the tags created in the EDB Console, in your cloud service provider's dashboard, with terraform, and other synced systems.

Reviewer: Based on the discussion in #dbaas-incubator, it seems like we want to lower the number of customer-added tags to 20? 

@amine-khemissi, @dhope-nagesh or @navnathedb  to confirm, please 🙏 

```suggestion
- In AWS, a resource can have a maximum number of 50 tags. The EDB Cloud Service and AWS apply some tags by default, so we recommend not adding more than 20 of your own tags. Take into account that the number of tags per resource is defined by the tags created in the EDB Console, in your cloud service provider's dashboard, with terraform, and other synced systems.
```

Reviewer 2: yeah, I think we can start with just 20 cc @marie-linet  @navnathedb

Reviewer 4: yes, 20 is a decent number to start with.

---

> title: "Tagging Cloud Service Providers"

Reviewer 2: The document is only valid for aws, not for the other cloud providers

Reviewer 5: should we point out somewhere that this feature is *currently* only available for AWS

Reviewer: What about 
```
Tagging AWS resources 
```

deepToC: true
---

> Tags are key-value pairs you can apply to resources created on Your Cloud Account for fine-grained access control and granular cost data at the bucket level.

Reviewer 2: ```suggestion
Tags are key-value pairs you can apply to the cloud provider resources created on Your Cloud Account for fine-grained access control and granular cost data.
```

Reviewer 2: resources that are tagged are :
- EC2 instances
- S3 buckets
- Loadbalencers
- Volumes

> Tags are created at the project level and managed through the Cloud Service, which is used across your cloud region.

Reviewer 2: what do we mean by Cloud Service ? 
and what do we mean by `used across you cloud region` ?

Reviewer: @amine-khemissi and @ivoryibuaka, I rephrased this part to make it more clear, let me know what you think. 

Cloud Service is the new name that the product team assigned to BigAnimal, and Console is the term to refer to the dashboard. I used Console here as I believe makes it more clear.

> Tags are created at the project level and managed through the Cloud Service, which is used across your cloud region.

Reviewer 5: i don't think i understand this sentence

## Create Tags

To create a tag:

1.  Log in to the Console.
    
1.  Go to the project's page.

1.  In the project's list, select your **Project**.

1.  On your project's home page left navigation, select **Cloud Providers**.

1.  On the cloud provider's page, select **Your Cloud Account**.

1.  On your cloud account tab, under **AWS**, select **Manage Tags**.

1.  Select **Add a Key Value pair**.

1.  Provide **Key** and **Value** and select **Save**.

1.  A **Confirm changes to AWS tags** message pops up.

1.  Select **Confirm** to add the tag.

> 1.  This **Key:Value** pair is added at your project level. It is inherited by all the resources in your project including regions and clusters.

Reviewer 2: not all resources, only four types of resources are tagged

This new tag appears in the list under **AWS** on Cloud Service Providers. 

> This tag is also added to the tags on your AWS Console. You can view this tag in the **Tags** tab of your EC2 instance resources on the AWS console.

Reviewer 2: ```suggestion
This tag is then propagated  on your AWS resources. For instance, you can view this tag in the **Tags** tab of your EC2 instance resources on the AWS console.
```

### Considerations

Consider following while adding a **Key:Value** pair:

- You can add a **Key:Value** pair or only **Key** in the **Key** field to add a tag.

- You cannot leave the **Key** field empty.

- The **Key** value must be unique within your project.

- The **Key** cannot have **aws** or **managedby** in the prefix.

- The maximum length for a **Key** is 127 characters and for **Value**, 255 characters.

- A resource can have a maximum number of 50 tags. The number of tags per resource is defined by the tags created in the EDB Console, in your cloud service provider's dashboard, with terraform, and other synced systems.

> 

Reviewer: @amine-khemissi I reworked this section, let me know what you think. I'd prefer keeping the word considerations instead of limitations, as I don't expect many customers encountering these issues, and limitations has a negative connotation, but I will bring this up with the docs team. 

@navnathedb let me know if there are other BA reserved word we should add here.

## Update Tags

To update a tag:

1.  Log in to the Console.

1.  Go to the project's page.

1.  In the project's list, select your **Project**.

1.  On your project's home page left navigation, select **Cloud Providers**.

1.  On the cloud provider's page, select **Your Cloud Account**.

1.  On your cloud account tab, under **AWS**, select **Manage Tags**.

1.  Update the **Value** field of any of the existing tags.

1.  Select **Save**.

1.  A **Confirm changes to AWS tags** message pops up.

1.  Select **Confirm** to update the tag.

> 1.  This **Key:Value** pair is updated at your project-level. The tag is inherited by all the resources in your project including regions and clusters.

Reviewer 2: see comment above

---
title: "Tagging AWS resources"
deepToC: true
redirects: 
  - /purl/cloudservice/csp_tagging
---

> Tags are key-value pairs you can apply to cloud provider resources created on your own cloud account, also known as [Your Cloud Account](/edb-postgres-ai/cloud-service/managed/). Resource tagging allows you to have fine-grained access control and granular cost data at the bucket level. When you add a tag, the following resources are labeled: 

Reviewer 2: Is it correct to talk about `granular cost data at the bucket level` ? because we tag other resources besides the buckets?

Reviewer 2: ```suggestion
Tags are key-value pairs you can apply to cloud provider resources created on your own cloud account, also known as [Your Cloud Account](/edb-postgres-ai/cloud-service/managed/). Resource tagging allows you to have fine-grained audit policy and cost control over your resources. When you add a tag, the following resources are labeled: 
```
